src/blockchain/Blockchain.ts

```diff
@@ -84,46 +99,160 @@ export default class Blockchain implements IBlockchainApi {
     identity: Identity,
     tx: SubmittableExtrinsic
   ): Promise<SubmittableResult> {
-    const accountAddress = identity.address
-    const nonce = await this.getNonce(accountAddress)
-    const signed: SubmittableExtrinsic = identity.signSubmittableExtrinsic(
-      tx,
-      nonce.toHex()
-    )
-    log.info(`Submitting ${tx.method}`)
-
-    return new Promise<SubmittableResult>((resolve, reject) => {
-      signed
-        .send((result) => {
-          log.info(`Got tx status '${result.status.type}'`)
-
-          if (ErrorHandler.extrinsicFailed(result)) {
-            log.warn(`Extrinsic execution failed`)
-            log.debug(`Transaction detail: ${JSON.stringify(result, null, 2)}`)
-            const extrinsicError: ExtrinsicError =
-              this.errorHandler.getExtrinsicError(result) || ERROR_UNKNOWN
-
-            log.warn(`Extrinsic error ocurred: ${extrinsicError}`)
-            reject(extrinsicError)
-          }
-          if (result.isFinalized) {
-            resolve(result)
-          } else if (result.isError) {
-            reject(
-              new Error(
-                `Transaction failed with status '${result.status.type}'`
+    try {
+      const nonce: Index = await this.getNonce(identity.address)
+      const signed = identity.signSubmittableExtrinsic(tx, nonce.toHex())
+      log.info(`Submitting ${tx.method} with Nonce ${nonce}`)
+      return new Promise<SubmittableResult>((resolve, reject) => {
+        signed
+          .send((result) => {
+            log.info(`Got tx status '${result.status.type}'`)
+            const { status } = result
+            if (ErrorHandler.extrinsicFailed(result)) {
+              log.warn(`Extrinsic execution failed`)
+              log.debug(
+                `Transaction detail: ${JSON.stringify(result, null, 2)}`
               )
-            )
-          }
-        })
-        .catch((err: Error) => {
-          // just reject with the original tx error from the chain
-          reject(err)
-        })
-    })
+              const extrinsicError: ExtrinsicError =
+                this.errorHandler.getExtrinsicError(result) || ERROR_UNKNOWN
+
+              log.warn(`Extrinsic error ocurred: ${extrinsicError}`)
+              this.resetAccountQueue(identity.address)
+              reject(extrinsicError)
+            }
+            if (result.isFinalized) {
+              this.resetAccountQueue(identity.address)
```

Comment:
the fact that you rely on the stuff that happens in the promise we return from submitTx could be a problem later on - we plan to expose the submittable extrinsic, so people are free to call `tx.submit()` and this would never be triggered

Reply:
Can we solve this any other way?